src/gui/sdlmain.cpp

```diff
@@ -519,84 +524,72 @@ static int int_log2 (int val) {
-{
-	// When the initial window is created in fullscreen (when requested
-	// window size matches the desktop, which is always the case for
-	// fullresolution=desktop), the position is forced to (0,0).
-	// In result, when leaving fullscreen the window is placed in top/left
-	// corner.
-	//                                                                        
-	// To work around this problem, center the window manually based on
-	// the original drawing size, and not window size.
-
-	if (fullscreen) {
-		const int x = (sdl.desktop.full.width - sdl.draw.width) / 2;
-		const int y = (sdl.desktop.full.height - sdl.draw.height) / 2;
-		return { x, y };
-	} else {
-		const int sdl_pos = SDL_WINDOWPOS_UNDEFINED_DISPLAY(sdl.displayNumber);
-		return { sdl_pos, sdl_pos };
-	}
-}
-
 static SDL_Window * GFX_SetSDLWindowMode(Bit16u width, Bit16u height, bool fullscreen, SCREEN_TYPES screenType)
 {
 	static SCREEN_TYPES lastType = SCREEN_SURFACE;
-	if (sdl.renderer) {
-		SDL_DestroyRenderer(sdl.renderer);
-		sdl.renderer = 0;
-	}
-	if (sdl.texture.pixelFormat) {
-		SDL_FreeFormat(sdl.texture.pixelFormat);
-		sdl.texture.pixelFormat = 0;
-	}
-	if (sdl.texture.texture) {
-		SDL_DestroyTexture(sdl.texture.texture);
-		sdl.texture.texture = 0;
-	}
-#if C_OPENGL
-	if (sdl.opengl.context) {
-		SDL_GL_DeleteContext(sdl.opengl.context);
-		sdl.opengl.context = 0;
-	}
-#endif
+
+	CleanupSDLResources();
+
 	int currWidth, currHeight;
 	if (sdl.window && sdl.resizing_window) {
 		SDL_GetWindowSize(sdl.window, &currWidth, &currHeight);
 		sdl.update_display_contents = ((width <= currWidth) && (height <= currHeight));
 		return sdl.window;
 	}
 
+	const bool first_window = !sdl.window;
+
 	/* If we change screen type, recreate the window. Furthermore, if
 	 * it is our very first time then we simply create a new window.
 	 */
-	if (!sdl.window || (lastType != screenType)) {
+	if (first_window || (lastType != screenType)) {
+
 		lastType = screenType;
 
 		if (sdl.window) {
 			SDL_DestroyWindow(sdl.window);
 		}
 
-		/* Don't create a fullscreen immediately. Reasons:
-		 * 1. This theoretically allows us to set window resolution and
-		 * then let SDL2 remember it for later (even if not actually done).
-		 * 2. It's a bit less glitchy to set a custom display mode for a
-		 * full screen, albeit it's still not perfect (at least on X11).
-		 */
-		const auto pos = get_default_pos(fullscreen);
-		const uint32_t flags = (screenType == SCREEN_OPENGL) ? SDL_WINDOW_OPENGL : 0;
-		sdl.window = SDL_CreateWindow("", pos.x, pos.y, width, height, flags);
+		uint32_t flags = 0;
+		if (screenType == SCREEN_OPENGL)
+			flags |= SDL_WINDOW_OPENGL;
+#if defined (WIN32)
+		// This is a hack for Windows 10 to prevent a crash in AMD OpenGL
+		// driver when window is being re-created by SDL2 internally to
+		// support OpenGL.
+		else if (screenType == SCREEN_TEXTURE && starts_with("opengl", sdl.render_driver))
+			flags |= SDL_WINDOW_OPENGL;
+#endif
```

Comment:
Is it possible to move this into the config-file code? 

For example, if this hack limits Windows systems to using a particular subset of the `output` and `texture_renderer` values, maybe we can move that limitation to the config-file code - which would make it impossible for Windows users to set bad combinations (and it would streamline this part).

Reply:
Not sure how it would work… the code depends on a runtime state and is needed only on windows. Right now Windows users cannot set bad combination - because of `starts_with("opengl", …)` the code handles `opengl`, `opengles`, `opengles2`, etc.

Hopefully I'll manage to figure out and prevent this hack in the future (somehow), so I would prefer not to propagate it to other source files.

Reply:
Thanks for explaining the dependencies; makes sense!